File: cplusplus/src/find_brackets.cpp
#include "find_brackets.hpp"

uMX find_closed_bracket(const std::string &src, uMX pos) 
{
    uMX openbt = 0;

    for (size_t i = pos + 1; i < src.size(); i++) {
        switch (src[i]) {
        case '[':
            openbt++;
            break;
        case ']':
            if (openbt == 0)
                return (uMX) i;
            else
                openbt--;
            break;
        
        default:
            break;
        }
    }
    return -1;
}


uMX find_opened_bracket(const std::string &src, uMX pos) 
{
    uMX closebt = 0;

    for (size_t i = pos - 1; i >= 0; i--) {
        switch (src[i]) {
        case ']':
            closebt++;
            break;
        case '[':
            if (closebt == 0)
                return (uMX) i;
            else
                closebt--;
            break;
        
        default:
            break;
        }
    }
    return -1;
}

File: rust/src/find_brackets.rs
pub fn find_closed_bracket(list: &Vec<char>, pos: usize) -> u32 {
    let mut openbt = 0;

    for i in (pos + 1)..list.len()
    {
        match list[i]
        {
            '[' =>  openbt += 1,

            ']' => {
                if openbt == 0 {
                    return i as u32;
                } else {
                    openbt -= 1;
                }
            }
            
            _ => {}
        }
    }

    u32::MAX
}

pub fn find_opened_bracket(list: &Vec<char>, pos: usize) -> u32 {
    let mut closebt = 0;

    for i in (0..(pos - 1)).rev()
    {
        match list[i]
        {
            ']' =>  closebt += 1,

            '[' => {
                if closebt == 0 {
                    return i as u32;
                } else {
                    closebt -= 1;
                }
            },
            
            _ => {}
        }
    }

    u32::MAX
}
